### PR TITLE
Introduce `onThrottledPointerMove` handler in `StateNode` and refactor the code to use that 

### PR DESCRIPTION
**Pull request by MitjaBezensek**
Abstract away the `run this logic on throttled pointer moves` into it's own event handler which handles the logic for subscribing to tick and also the logic for marking things dirty when pointer move is triggered. The [main change](https://github.com/tldraw/tldraw/pull/3107/files#diff-470d2a569509456407ff57b3b01974a4ff6d291a76855f3986d3650ce5c77ff1) is in the `StateNode` where we now:
- Introduce `isDirty` field and manage its value: we set it to `false` when entering / exiting the node, after we call the `onThrottledPointerMove` handler, and we set it to `true` when the `pointerMove` event happens. 
- We call `onThrottledPointerMove` on tick if the `isDirty` flag is set.

Marking things as dirty now only works for the `onPointerMove` handler so you might need to manually set the flag if things have changed outside of that - like the key events in the `Drawing`.

Follow up PR to #2977 based on [this comment](https://github.com/tldraw/tldraw/pull/2977#pullrequestreview-1927555665).

### Change Type

- [ ] `patch` — Bug fix
- [x] `minor` — New feature
- [ ] `major` — Breaking change
- [ ] `dependencies` — Changes to package dependencies[^1]
- [ ] `documentation` — Changes to the documentation only[^2]
- [ ] `tests` — Changes to any test code only[^2]
- [ ] `internal` — Any other changes that don't affect the published package[^2]
- [ ] I don't know

[^1]: publishes a `patch` release, for devDependencies use `internal`
[^2]: will not publish a new version

### Release Notes

- Introduces a new `onThrottledPointerMove` event handler, which can be used for computation heavy pointer moves which you might want to throttle.

### CODE REVIEW

File: packages/editor/src/lib/editor/tools/StateNode.ts [hunk 1]
```diff
@@ -73,6 +74,8 @@ export abstract class StateNode implements Partial<TLEventHandlers> {
 	initial?: string
 	children?: Record<string, StateNode>
 	parent: StateNode
+	// Flag that indicates whether the `StateNode` contains some changes that have not been flushed to the Editor.
+	isDirty: boolean
```
Comment (ds300) [inline packages/editor/src/lib/editor/tools/StateNode.ts:78]: I'm a bit wary of making this public. Seems like an implementation detail and I think I'd rather keep it private or internal and related to pointer movement specifically for now until we see a wider pattern develop.

**Comment by MitjaBezensek** [comment]
> bit of an API design Q here, but can you talk a bit about this being a separate event rather than the default behaviour? i'm sort of wondering if there are things we can do to help devs do the right thing by default here, and if throttling is the answer it might be nice for that to be a bit more of a built-in concept. the browsers own [`PointerEvent.getCoalescedEvents()` api](https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent/getCoalescedEvents) might be a good starting point if we did want to look into making throttling be the default pointer move behaviour

I think the default behaviour for most operation should be to use the regular `onPointerMove` event. Going to the throttled versions introduces some complexity, which developers probably don't need. So it should be used when performance becomes a problem. You still probably want to use the `onPointerMove` on top of the throttled version. You do that in order to capture all the inputs in `onPointerMove` (increased granularity), but only throttle the state updates. 

I see that `getCoalescedEvents` gives you access to un-coalesced mouse positions, but you might also be interested in some other events. Like what happens if the user pressed Shift between two coalesced events, would we be able to handle that?

File: packages/tldraw/src/lib/shapes/draw/toolStates/Drawing.ts [hunk 2]
```diff
@@ -324,6 +319,7 @@ export class Drawing extends StateNode {
 
 		const newPoint = { x, y, z: this.isPen ? +(z! * 1.25).toFixed(2) : 0.5 }
 
+		// We also set the dirty flag here, since we might reach this from key events
 		this.isDirty = true
```
[the remark this answers is not in the document]
Comment (MitjaBezensek) [reply packages/tldraw/src/lib/shapes/draw/toolStates/Drawing.ts:323]: We would incorrectly handle events like pressing / releasing the `Shift` key. Right now we only automatically mark the state as dirty on `pointerMove` events, but there are some other events that also cause the state to change. Without this line some tests that involve shift fail - think it was tests that check how snapping to straight segments works when shift is held.

Think me and Steve decided it's better to put this into `processUpdates` function, as it is the main function that handles the updates. Alternately we could also put this into `onKeyDown` and `onKeyUp` that call `processUpdates`.

File: packages/editor/src/lib/editor/tools/StateNode.ts [hunk 3]
```diff
@@ -148,6 +151,7 @@ export abstract class StateNode implements Partial<TLEventHandlers> {
 	handleEvent = (info: Exclude<TLEventInfo, TLPinchEventInfo>) => {
 		const cbName = EVENT_NAME_MAP[info.name]
 		const x = this.getCurrent()
+		if (cbName === 'onPointerMove') this.isDirty = true
 		this[cbName]?.(info as any)
```
[the remark this answers is not in the document]
Comment (MitjaBezensek) [reply packages/editor/src/lib/editor/tools/StateNode.ts:155]: I guess that's to immediately flush the changes?

**Comment by MitjaBezensek** [comment]
Decided not to go with this. We might want to introduce something like this in the future, but for right now I prefer some repetition in the `StateNodes` over exposing this API.

**Pull request closed without merging**